ccl/sqlproxyccl: do not classify TCP probes as a connection error

jaylim-crl opened the pull request:
#### ccl/sqlproxyccl: do not classify TCP probes as a connection error 

Previously, if a client attempted to open a TCP connection without any packets
sent, the proxy will log an error (i.e. while receiving startup message), and
close the connection. This scenario is common with TCP probes, and this is
what cloud load balancers generally do. To avoid these probes spamming the
proxy logs, we will silently ignore these class of errors, and this commit
does that.
  
#### ccl/sqlproxyccl: add proxy.sql.accepted_conns metric to the proxy 

This commit adds a new proxy.sql.accepted_conns metric to indicate the number
of accepted connections by the proxy. The previous commit removed logging
messages for connections that do not have any data packets. This new metric
would aid in debugging (e.g. high cpu usage).

Release note: None

Epic: none

jaylim-crl committed:
ccl/sqlproxyccl: do not classify TCP probes as a connection error

Previously, if a client attempted to open a TCP connection without any packets
sent, the proxy will log an error (i.e. while receiving startup message), and
close the connection. This scenario is common with TCP probes, and this is
what cloud load balancers generally do. To avoid these probes spamming the
proxy logs, we will silently ignore these class of errors, and this commit
does that.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98279)
<!-- Reviewable:end -->

jaylim-crl commented:
I've tested that this change works well with the NLB in IP-mode, and healthchecks are configured to point to the exposed SQL port, as desired.

pjtatlow commented on pkg/ccl/sqlproxyccl/proxy_handler_test.go:
```diff
@@ -657,6 +657,24 @@ func TestProxyRefuseConn(t *testing.T) {
 	require.Equal(t, int64(0), s.metrics.AuthFailedCount.Count())
 }
 
+func TestProxyTCPProbe(t *testing.T) {
```
I could be missing something, but I'm not sure what this test is asserting actually. I think this test would likely pass even without these changes. Is there a way we can assert on how the `handle` method treats the connection closing without sending anything? If not, that's fine, but I don't see what this test does except check that we can accept TCP connections, but the other tests do that already.

pjtatlow reviewed (approved):
LGTM! This will definitely make things a lot nicer with the TCP health checks.

JeffSwenson commented:
The overall change looks good to me. Do these requests increment any of the connection metrics? We expect the traffic from load balancers, but it could also signal something like a DOS attack.

jaylim-crl committed:
ccl/sqlproxyccl: add proxy.sql.accepted_conns metric to the proxy

This commit adds a new proxy.sql.accepted_conns metric to indicate the number
of accepted connections by the proxy. The previous commit removed logging
messages for connections that do not have any data packets. This new metric
would aid in debugging (e.g. high cpu usage).

Release note: None

Epic: none

jaylim-crl reviewed (commented):
TFTR!

> We may want an "healthcheck_connection" metric that directly measures this.

Done. I added `proxy.sql.accepted_conns` to count the number of accepted connections.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98279#-:-NQ6NGJhDf23f4DFub2O:bb717qg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson and @pjtatlow)

___
*[`pkg/ccl/sqlproxyccl/proxy_handler_test.go` line 660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98279#-NQ3EdPtVpzeEq_IxXoh-r1-660:-NQ6NNMR7sNG46cUg7dz:b-2rym63) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca66d7ad173728e98b65b6ca6d32c145aeefcef/pkg/ccl/sqlproxyccl/proxy_handler_test.go#L660)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I’m going to see if I can update the test to check on handle directly.
</blockquote></details>

Done. Made the test call `handle` directly.


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

pjtatlow reviewed (approved):
LGTM!

jaylim-crl commented:
TFTRs!

bors r=pjtatlow,JeffSwenson
